doc/macro_reference.qbk

```diff
@@ -1081,6 +1082,33 @@ that are not yet supported by a particular compiler or library.
 
 [endsect]
 
+[section Macros that allow C++20 features to be used in previous standard versions]
+
+[table
+[[macro][Description]]
+[[`BOOST_IS_CONSTANT_EVALUATED()`][Evaluated to `true` when the current context is `constexpr` and false otherwise.
+                                 This macro may be functional in C++14 onwards if the compiler has a suitable intrinsic we recognise (MSVC/Clang/GCC are currently supported).
+
+                                 Note that before using this macro, the program must include `<type_traits>`.
+
+                                 The macro always evaluates to `false` on unsupported/old compilers.]]
+[[`BOOST_IS_CONSTANT_EVALUATED_INT(integer-expression)`][Evaluates to `true` when `integer-expression` is `constexpr`.
+                                 The macro may be functional in C++14 onwards if the compiler has a suitable intrinsic, in particular this macro is functional for gcc-6 and onwards
+                                 as well as for compilers which have direct support for `std::is_constant_evaluated`.
+
+                                 Note that before using this macro, the program must include `<type_traits>`.
+
+                                 Note that this macro may behave subtly different from `std::is_constant_evaluated()` in that it may evaluate to
+                                 `true` when code is evaluated in a non-constexpr context, but is none the less evaluated at "compile time" as a result
+                                 of program optimisation.
+
+                                 The macro always evaluates to `false` on unsupported/old compilers.]]
```

Comment:
Nod.  There will certainly be tricky cases where you can abuse it into doing the wrong thing, if it's easier I could just omit that macro.

Reply:
>  if it's easier I could just omit that macro.

Hard for me to judge.
My suggestion is to ask the users what they feel. I would personally never use it since I consider it too tricky, but I am not really an expert library developer and I have no usecases for it.

Reply:
`__builtin_constant_p(n)` is not a constant expression, and it does not indicate whether `n` is a constant expression either. See [here](https://gcc.godbolt.org/z/3h1KP9oK8). What it shows is whether the compiler is able to deduce the value of `n` at compile time, including as a result of optimizations (e.g. inlining). This means that its value may change depending on optimization level used. If it returns `true`, `n` may not be a constant expression in C++ definition and thus may not be used in C++ constant expressions.

I think `BOOST_IS_CONSTANT_EVALUATED_INT` needs to be renamed to avoid confusion with `std::is_constant_evaluated`. `__builtin_constant_p` is useful, but it is a very different thing.